internal/certificates/certificates.go

```diff
@@ -0,0 +1,126 @@
+package certificates
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"os"
+	"sync"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/xerrors"
+)
+
+var (
+	// fileCache
+	//  map[string][]*x509.Certificate
+	fileCache sync.Map
+	// FileCacheEnabled turns caching in ParseCertificatesFromFile on or off.
+	// This varbialbe MUST not be set concurrently with calls to ParseCertificatesFromFile.
+	FileCacheEnabled = true
+	// FileCacheHook (if not nil) is called on every call to ParseCertificatesFromFile
+	// if FileCacheEnabled = true. Its argument tells whether there was a cache hit.
+	// This varbialbe MUST not be set concurrently with calls to ParseCertificatesFromFile.
+	FileCacheHook func(isHit bool)
+)
+
+// ParseCertificatesFromFile reads and parses pem-encoded certificate(s) from file.
+func ParseCertificatesFromFile(file string) ([]*x509.Certificate, error) {
+	if !FileCacheEnabled {
+		certs, err := parseCertificatesFromFile(file)
+		return certs, xerrors.WithStackTrace(err)
+	}
+
+	value, exists := fileCache.Load(file)
```

Comment:
Yes, the problem is possible. We have another cache `string(decoded pem block) -> *x509.Certificate` which seems like what you're talking about. Caching by filename saves some resources compared to only PEM cache (as indicated by benchmark).

Possible solutions (as I see them) are:

 - remove caching by filename at all and leave only PEM cache
 - delete filename cache records on come condition (e.g. on certificate expiration time `cert.NotAfter`) + maybe give users an option of manually forcing uncached certificate loading

Caching by **file** contents / hash is not good because the most expensive operation is `os.ReadFile`. I don't have a benchmark testing this kind of caching, but I did compare `os.ReadFile` with `pem.Decode`, and the former is much more expensive.

Reply:
We could also use `os.Lstat()` and `fileInfo.ModTime()` to determine if the file was modified after caching. Here is a comparison between `os.ReadFile` and `os.Lstat`:

```
BenchmarkFileStuff/os.ReadFile-8         	  341007	      3507 ns/op	    1080 B/op	       5 allocs/op
BenchmarkFileStuff/os.Lstat___-8         	 1000000	      1081 ns/op	     272 B/op	       2 allocs/op
```

Reply:
In your example you can read file 285000 times per second.
How often do you read certificates in your case?

Reply:
Currently filename cache saves about 4.6 us and 2.3 KB per child connection (call to `db.With`):

```
BenchmarkWithFileCache/no_cache-8                  37642             32658 ns/op           44066 B/op        247 allocs/op
BenchmarkWithFileCache/cert_cache-8                43017             27734 ns/op           41731 B/op        210 allocs/op
BenchmarkWithFileCache/both_caches-8               52866             23121 ns/op           39379 B/op        191 allocs/op
```

So I think that expensiveness of `ReadFile`/`Lstat` should be compared to these values, not just by itself.

Unfortunately, I don't have enough info on real-life cases related to the issue. Maybe, the original issuer might provide some details.